service/matching/version_rule_helpers.go

```diff
@@ -0,0 +1,244 @@
+package matching
+
+import (
+	"fmt"
+	"slices"
+	"time"
+
+	"go.temporal.io/api/serviceerror"
+	"go.temporal.io/api/taskqueue/v1"
+	"go.temporal.io/api/workflowservice/v1"
+
+	persistencepb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/common"
+	hlc "go.temporal.io/server/common/clock/hybrid_logical_clock"
+	"go.temporal.io/server/common/util"
+)
+
+// checkAssignmentConditions returns an error if the new set of assignment rules don't meet the following requirements:
+// - No more rules than dynamicconfig.VersionAssignmentRuleLimitPerQueue
+// - Once assignment rules have been added, there must always be at least one "unfiltered" assignment rule, which can accept any build id
+// - No assignment rule's TargetBuildId can be a member of an existing version set
+func checkAssignmentConditions(g *persistencepb.VersioningData, maxARs int, force bool) error {
+	if force == true {
+		return nil
+	}
+	rules := slices.Clone(g.GetAssignmentRules())
+	if maxARs > 0 && len(rules) > maxARs {
+		// todo carly: should the max # of rules count rules that have been deleted but not yet cleaned up from the db?
+		return serviceerror.NewFailedPrecondition(fmt.Sprintf("update exceeds number of assignment rules permitted in namespace dynamic config (%v/%v)", len(rules), maxARs))
+	}
+	foundUnfiltered := false
+	for _, dbRule := range rules {
+		rule := dbRule.GetRule()
+		if dbRule.DeleteTimestamp == nil && isUnfiltered(rule) {
+			foundUnfiltered = true
+		}
+		tbid := rule.GetTargetBuildId()
+		if dbRule.DeleteTimestamp == nil && isInVersionSet(tbid, g.GetVersionSets()) {
+			return serviceerror.NewFailedPrecondition(fmt.Sprintf("update breaks requirement, target build id %s is already a member of version set", tbid))
+		}
+	}
+	if !foundUnfiltered {
+		return serviceerror.NewFailedPrecondition("update breaks requirement that at least one assignment rule must have no ramp or hint")
+	}
+	return nil
+}
+
+// isUnfiltered returns true if the assignment rule has no hint, no worker ratio ramp, and no ramp percentage, or a ramp percentage of 100
+func isUnfiltered(ar *taskqueue.BuildIdAssignmentRule) bool {
+	percentageRamp := ar.GetPercentageRamp()
+	return ar.GetHintFilter() == "" &&
+		ar.GetWorkerRatioRamp() == nil &&
+		(percentageRamp == nil || (percentageRamp != nil && percentageRamp.RampPercentage == 100))
```

Comment:
maybe it's cleaner to assume the percentage cannot be 100%. For user's ease of use, CLI can accept 100% but remove such a ramp message from the proto when it calls the server.